Add test that checks the 2-pass AST matching works

Thoemi09 opened the pull request:
- without 2-pass, the added test will not work

Thoemi09 committed:
Add test checks that the 2-pass AST matching works

- without 2-pass, the added test will not work

parcollet commented:
* Your PR must include the cxx changes, or the non clair builds will fail
*[minor]  try to use use zzfoo or names which are later in the alphatical order, it will minimize the cxx diff.

Thoemi09 committed:
Implement suggestions to make 2-pass test cleaner

- add ctor to avoid cluttering of generated code
- rename functions to make changes in generated code minimal

parcollet committed:
Clarify comments on two-pass AST matcher

Updated comments to clarify the need for two-pass AST matching.
